File: android/src/main/java/com/monri/ionic/IonicMonriPlugin.java
package com.monri.ionic;

import android.content.Context;
import android.content.Intent;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.monri.android.Monri;
import com.monri.android.model.Card;
import com.monri.android.model.ConfirmPaymentParams;
import com.monri.android.model.CustomerParams;
import com.monri.android.model.MonriApiOptions;
import com.monri.android.model.SavedCard;
import com.monri.android.model.TransactionParams;

import org.json.JSONException;
import org.json.JSONObject;

import java.lang.ref.WeakReference;

@CapacitorPlugin(name = "IonicMonri")
public class IonicMonriPlugin extends Plugin {
    private Monri monri;
    private String savedPluginId;

    @PluginMethod
    public void confirmPayment(PluginCall call) {

        if (!call.hasOption("params")) {
            call.reject("missing params object");
            return;
        }

        final boolean doesContainValidPaymentMethod = call.getObject("params").has("card") ||
                call.getObject("params").has("savedCard");

        if (!doesContainValidPaymentMethod) {
            call.reject("Unsupported payment method, 'card' or 'savedCard' not found");
            return;
        }

        getBridge().saveCall(call);
        savedPluginId = call.getCallbackId();

        MonriApiOptions monriApiOptions = parseMonriApiOptions(call);
        ConfirmPaymentParams confirmPaymentParams = parseConfirmPaymentParams(call);

        IonicMonriPlugin.writeMetaData(getContext(), String.format("Android-SDK:Ionic:%s", BuildConfig.MONRI_IONIC_PLUGIN_VERSION));

        monri = new Monri(getActivity(), monriApiOptions);
        monri.confirmPayment(getActivity(), confirmPaymentParams);

    }

    public void monriHandleOnActivityResult(final int requestCode, final int resultCode, final Intent data) {
        final boolean monriPaymentResult = monri.onPaymentResult(
                requestCode,
                data,
                new PaymentResultCallback(
                        new WeakReference<>(getContext()),
                        new WeakReference<>(getBridge()),
                        savedPluginId
                ));
    }

    private ConfirmPaymentParams parseConfirmPaymentParams(final PluginCall params) {

        final JSObject paramsObject = params.getObject("params");
        final String clientSecret = paramsObject.getString("clientSecret");

        Card card = null;
        SavedCard savedCard = null;

        if (params.getObject("params").has("card")) {
            final JSObject cardJSObject = paramsObject.getJSObject("card");
            card = new Card(
                    getNullableString(cardJSObject, "pan"),
                    Integer.valueOf(getNullableString(cardJSObject, "expiryMonth")),
                    Integer.valueOf(getNullableString(cardJSObject, "expiryYear")),
                    getNullableString(cardJSObject, "cvv")
            );

            card.setTokenizePan(cardJSObject.getBoolean("saveCard",false));

        } else if (params.getObject("params").has("savedCard")) {
            final JSObject cardJSObject = paramsObject.getJSObject("savedCard");
            savedCard = new SavedCard(
                    getNullableString(cardJSObject, "panToken"),
                    getNullableString(cardJSObject, "cvv")
            );

        } else {
            throw new IllegalStateException("Unsupported payment method, 'card' or 'savedCard' not found");
        }

        final JSObject transactionJSObject = paramsObject.getJSObject("transaction");
        final CustomerParams customerParams = new CustomerParams()
                .setAddress(getNullableString(transactionJSObject, "address"))
                .setFullName(getNullableString(transactionJSObject, "fullName"))
                .setCity(getNullableString(transactionJSObject, "city"))
                .setZip(getNullableString(transactionJSObject, "zip"))
                .setPhone(getNullableString(transactionJSObject, "phone"))
                .setCountry(getNullableString(transactionJSObject, "country"))
                .setEmail(getNullableString(transactionJSObject, "email"));

        ConfirmPaymentParams confirmPaymentParams = ConfirmPaymentParams.create(
                clientSecret,
                card == null ? savedCard.toPaymentMethodParams() : card.toPaymentMethodParams(),
                new TransactionParams()
                        .set("order_info", getNullableString(transactionJSObject, "orderInfo"))
                        .set(customerParams)
        );

        return confirmPaymentParams;
    }

    private MonriApiOptions parseMonriApiOptions(final PluginCall options) {
        final JSObject object = options.getObject("options");

        String authenticityToken = getNullableString(object, "authenticityToken");
        boolean developmentMode = object.getBool("developmentMode");
        MonriApiOptions monriApiOptions = new MonriApiOptions(authenticityToken, developmentMode);
        return monriApiOptions;
    }

    private String getNullableString(
            JSObject jsonObject,
            String key
    ) {
        return jsonObject.getString(key);

    }

    private static void writeMetaData(Context context, String library) {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        sharedPreferences.edit().putString("com.monri.meta.library", library).apply();
    }

}

File: ios/Plugin/IonicMonriPlugin.swift
import Foundation
import Capacitor
import Monri

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(IonicMonriPlugin)
public class IonicMonriPlugin: CAPPlugin {

    @objc func confirmPayment(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let module = self else {
                return
            }

            module.__call(call)
        }
    }

    private func __call(_ call: CAPPluginCall){
        if(!isPaymentMethodSupported(call)){
            return
        }

        do {
            let options = try parseMonriApiOptions(call)
            let confirmPaymentParams = try parseConfirmPaymentParams(call)
            let delegate = UIApplication.shared.delegate!
            let vc = delegate.window!!.rootViewController!
            let monri = MonriApi(vc, options: options)

            writeMetaData()

            monri.confirmPayment(confirmPaymentParams) { [weak self] result in
                switch (result) {
                case .result(let paymentResult):
                    var rv: [String: Any] = [
                        "status": paymentResult.status,
                        "currency": paymentResult.currency,
                        "amount": paymentResult.amount,
                        "orderNumber": paymentResult.orderNumber,
                        "createdAt": paymentResult.createdAt,
                        "transactionType": paymentResult.transactionType,
                    ]

                    if let pm = paymentResult.paymentMethod {
                        rv["paymentMethod"] = [
                            "type": pm.type,
                            "data": [
                                "brand": pm.data["brand"]!,
                                "expirationDate": pm.data["expiration_date"]!,
                                "issuer": pm.data["issuer"]!,
                                "masked": pm.data["masked"]!,
                                "token": pm.data["token"]!
                            ]
                        ]
                    }

                    if let panToken = paymentResult.panToken {
                        rv["panToken"] = panToken
                    }

                    rv["errors"] = paymentResult.errors

                    call.resolve(rv)
                case .error(let e):
                    call.resolve(["status": "error", "errors": [e.localizedDescription]])
                case .declined(let d):
                    call.resolve(["status": d.status])
                case .pending:
                    call.resolve(["status": "pending"])
                }
            }


        } catch {
            if let configurationError = error as? MonriAndroidIosConfirmPaymentError {
                switch (configurationError) {
                case .configurationError(let m):
                    call.reject(MonriAndroidIosConfirmPaymentErrorCodes.configurationError.rawValue)
                case .parsingError(let m):
                    call.reject(MonriAndroidIosConfirmPaymentErrorCodes.parsingError.rawValue, m, error)
                case .failedToParseMonriApiOptions:
                    call.reject(MonriAndroidIosConfirmPaymentErrorCodes.failedToParseMonriApiOptions.rawValue, "Failed to parse api options", error)
                case .missingRequiredAttribute(let m):
                    call.reject(MonriAndroidIosConfirmPaymentErrorCodes.missingRequiredAttribute.rawValue, m, error)
                }

            } else {
                call.reject(MonriAndroidIosConfirmPaymentErrorCodes.unknown.rawValue, error.localizedDescription, error)
            }
        }
    }

     private func writeMetaData(){
            let version: String = Bundle(identifier: "org.cocoapods.IonicMonriAndroidIos")?.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"

            let defaults = UserDefaults.standard
            defaults.set("iOS-SDK:Ionic:\(version)", forKey: "com.monri.meta.library")
        }

    private func isPaymentMethodSupported(_ call: CAPPluginCall) -> Bool{
        let paramsObject: JSObject? = call.getObject("params")
        if(paramsObject == nil){
            call.reject("missing params object");
            return false
        }

        let savedCardJSObject: JSObject? = paramsObject!["savedCard"] as? JSObject
        let cardJSObject: JSObject? = paramsObject!["card"] as? JSObject

        let doesContainValidPaymentMethod: Bool = (savedCardJSObject != nil || cardJSObject != nil)

        if (!doesContainValidPaymentMethod) {
            call.reject("Unsupported payment method, 'card' or 'savedCard' not found");
        }
        return doesContainValidPaymentMethod;
    }

    private func parseMonriApiOptions(_ call: CAPPluginCall) throws -> MonriApiOptions {
        guard let monriApiOptionsJSObject: JSObject = call.getObject("options") else {
            throw MonriAndroidIosConfirmPaymentError.missingRequiredAttribute("params")
        }

        guard let authenticityToken = monriApiOptionsJSObject["authenticityToken"] as? String else {
            throw MonriAndroidIosConfirmPaymentError.missingRequiredAttribute("authenticityToken")
        }


        let developmentMode = (monriApiOptionsJSObject["developmentMode"] as? NSNumber) != 0


        return MonriApiOptions(authenticityToken: authenticityToken, developmentMode: developmentMode)
    }

    private func parseConfirmPaymentParams(_ call: CAPPluginCall) throws -> ConfirmPaymentParams {

        guard let paramsObject: JSObject = call.getObject("params") else {
            throw MonriAndroidIosConfirmPaymentError.missingRequiredAttribute("params")
        }

        guard let clientSecret: String = paramsObject["clientSecret"] as? String else {
            throw MonriAndroidIosConfirmPaymentError.missingRequiredAttribute("clientSecret")
        }

        let savedCardJSObject: JSObject? = paramsObject["savedCard"] as? JSObject
        let cardJSObject: JSObject? = paramsObject["card"] as? JSObject

        if(savedCardJSObject == nil && cardJSObject == nil){
            throw MonriAndroidIosConfirmPaymentError.missingRequiredAttribute("card or savedCard")
        }

        let paymentMethod = cardJSObject == nil ? SavedCard(
            panToken: try requiredStringAttribute(savedCardJSObject!, "panToken", "params.savedCard.panToken"),
            cvc: try requiredStringAttribute(savedCardJSObject!, "cvv", "params.savedCard.cvv")
        ).toPaymentMethodParams()
        :Card(
            number: try requiredStringAttribute(cardJSObject!, "pan", "params.card.pan"),
            cvc: try requiredStringAttribute(cardJSObject!, "cvv", "params.card.cvv"),
            expMonth: try requiredIntAttribute(cardJSObject!, "expiryMonth", "params.card.expiryMonth"),
            expYear: try requiredIntAttribute(cardJSObject!, "expiryYear", "params.card.expiryYear"),
            tokenizePan: (cardJSObject!["saveCard"] as? Bool) ?? false
        ).toPaymentMethodParams()

        guard let transactionJSObject: JSObject = paramsObject["transaction"] as? JSObject else {
            throw MonriAndroidIosConfirmPaymentError.missingRequiredAttribute("transaction")

        }

        let customerParams = CustomerParams(email: getString(transactionJSObject, "email"),
                                            fullName: getString(transactionJSObject, "fullName"),
                                            address: getString(transactionJSObject, "address"),
                                            city: getString(transactionJSObject, "city"),
                                            zip: getString(transactionJSObject, "zip"),
                                            phone: getString(transactionJSObject, "phone"),
                                            country: getString(transactionJSObject, "country")
        )

        return ConfirmPaymentParams(paymentId: clientSecret, paymentMethod: paymentMethod, transaction: TransactionParams.create()
                                        .set(customerParams: customerParams)
                                        .set("order_info", transactionJSObject["orderInfo"] as? String)
        )
    }


    private func requiredStringAttribute(_ jsObject: JSObject, _ key: String, _ path: String? = nil) throws -> String {
        guard let value = jsObject[key] as? String else {
            throw MonriAndroidIosConfirmPaymentError.missingRequiredAttribute(path ?? key)
        }

        return value
    }

    private func getString(_ jsObject: JSObject, _ key: String) -> String {
        return jsObject[key] as! String
    }

    private func requiredIntAttribute(_ jsObject: JSObject, _ key: String, _ path: String? = nil) throws -> Int {
        guard let value = jsObject[key] as? Int else {
            throw MonriAndroidIosConfirmPaymentError.missingRequiredAttribute(path ?? key)
        }

        return value
    }

    enum MonriAndroidIosConfirmPaymentErrorCodes: String {
        case failedToParseMonriApiOptions
        case configurationError
        case parsingError
        case missingRequiredAttribute
        case unknown
    }

    enum MonriAndroidIosConfirmPaymentError: Error {
        case failedToParseMonriApiOptions
        case configurationError(String)
        case parsingError(String)
        case missingRequiredAttribute(String)
    }

}
